File: src/main/java/com/fengsheng/skill/TouTian.java
package com.fengsheng.skill;

import com.fengsheng.*;
import com.fengsheng.card.JieHuo;
import com.fengsheng.phase.FightPhaseIdle;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Role;
import com.google.protobuf.GeneratedMessageV3;
import org.apache.log4j.Logger;

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

/**
 * 鄭文先技能【偷天】：争夺阶段你可以翻开此角色牌，然后视为你使用了一张【截获】。
 */
public class TouTian extends AbstractSkill {
    private static final Logger log = Logger.getLogger(TouTian.class);

    @Override
    public void init(Game g) {

    }

    @Override
    public SkillId getSkillId() {
        return SkillId.TOU_TIAN;
    }

    @Override
    public ResolveResult execute(Game g) {
        return null;
    }

    @Override
    public void executeProtocol(Game g, Player r, GeneratedMessageV3 message) {
        if (!JieHuo.canUse(g, r)) return;
        if (r.isRoleFaceUp()) {
            log.error("你现在正面朝上，不能发动[偷天]");
            return;
        }
        var pb = (Role.skill_tou_tian_tos) message;
        if ((r instanceof HumanPlayer humanPlayer) && !humanPlayer.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + humanPlayer.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        r.incrSeq();
        r.addSkillUseCount(getSkillId());
        g.playerSetRoleFaceUp(r, true);
        log.info(r + "发动了[偷天]");
        for (Player p : g.getPlayers()) {
            if (p instanceof HumanPlayer player)
                player.send(Role.skill_tou_tian_toc.newBuilder().setPlayerId(player.getAlternativeLocation(r.location())).build());
        }
        JieHuo.execute(null, g, r);
    }

    public static boolean ai(FightPhaseIdle e, final Skill skill) {
        if (e.whoseFightTurn.isRoleFaceUp())
            return false;
        Player player = e.whoseFightTurn;
        var colors = e.messageCard.getColors();
        if (e.inFrontOfWhom == player || (e.isMessageCardFaceUp || player == e.whoseTurn) && colors.size() == 1 && colors.get(0) == Common.color.Black)
            return false;
        if (ThreadLocalRandom.current().nextBoolean())
            return false;
        GameExecutor.post(e.whoseFightTurn.getGame(), () -> skill.executeProtocol(
                e.whoseFightTurn.getGame(), e.whoseFightTurn, Role.skill_tou_tian_tos.getDefaultInstance()
        ), 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/kotlin/skill/TouTian.kt
package com.fengsheng.skill

import com.fengsheng.*
import com.fengsheng.card.JieHuo
import com.fengsheng.phase.FightPhaseIdle
import com.fengsheng.protos.Common.card_type.Jie_Huo
import com.fengsheng.protos.Common.card_type.Wu_Dao
import com.fengsheng.protos.Role.skill_tou_tian_tos
import com.fengsheng.protos.skillTouTianToc
import com.fengsheng.protos.skillTouTianTos
import com.fengsheng.skill.SkillId.TOU_TIAN
import com.google.protobuf.GeneratedMessage
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit

/**
 * 鄭文先技能【偷天】：争夺阶段你可以翻开此角色牌，然后视为你使用了一张【截获】。
 */
class TouTian : ActiveSkill {
    override val skillId = TOU_TIAN

    override val isInitialSkill = true

    override fun canUse(fightPhase: FightPhaseIdle, r: Player): Boolean = !r.roleFaceUp

    override fun executeProtocol(g: Game, r: Player, message: GeneratedMessage) {
        if (!JieHuo.canUse(g, r)) return
        if (r.roleFaceUp) {
            logger.error("你现在正面朝上，不能发动[偷天]")
            r.sendErrorMessage("你现在正面朝上，不能发动[偷天]")
            return
        }
        val pb = message as skill_tou_tian_tos
        if (r is HumanPlayer && !r.checkSeq(pb.seq)) {
            logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        r.incrSeq()
        r.addSkillUseCount(skillId)
        g.playerSetRoleFaceUp(r, true)
        logger.info("${r}发动了[偷天]")
        g.players.send { skillTouTianToc { playerId = it.getAlternativeLocation(r.location) } }
        JieHuo.execute(null, g, r)
    }

    companion object {
        fun ai(e: FightPhaseIdle, skill: ActiveSkill): Boolean {
            val player = e.whoseFightTurn
            !player.roleFaceUp || return false
            !player.game!!.isEarly || player.game!!.players.anyoneWillWinOrDie(e) || return false
            e.inFrontOfWhom !== player || return false
            val oldValue = player.calculateMessageCardValue(e.whoseTurn, e.inFrontOfWhom, e.messageCard)
            val newValue = player.calculateMessageCardValue(e.whoseTurn, player, e.messageCard)
            newValue > oldValue || return false
            val result = player.calFightPhase(e)
            if (result != null && result.cardType in listOf(Jie_Huo, Wu_Dao) && result.value >= newValue) return false
            GameExecutor.post(e.whoseFightTurn.game!!, {
                skill.executeProtocol(e.whoseFightTurn.game!!, e.whoseFightTurn, skillTouTianTos { })
            }, 3, TimeUnit.SECONDS)
            return true
        }
    }
}